[Site Isolation] Main frame's Document.activeElement incorrectly reports iframe after focus is relinquished from a cross-origin iframe to the user agent chrome

aprotyas opened the pull request:
#### 5493989ddb66a090c5919f396544342e4768226b
<pre>
[Site Isolation] Main frame&apos;s Document.activeElement incorrectly reports iframe after focus is relinquished from a cross-origin iframe to the user agent chrome
<a href="https://bugs.webkit.org/show_bug.cgi?id=300154">https://bugs.webkit.org/show_bug.cgi?id=300154</a>
<a href="https://rdar.apple.com/161863002">rdar://161863002</a>

Reviewed by Aditya Keerthi.

When focus is relinquished from a cross-origin iframe to the user agent
chrome (e.g., an AppKit text field), document.activeElement in the main
frame incorrectly still references the iframe element. This occurs
because broadcastFocusedFrameToOtherProcesses() only supported
broadcasting a valid FrameIdentifier, not nullptr when focus moves to
chrome.

The fix changes broadcastFocusedFrameToOtherProcesses() to accept
std::optional&lt;FrameIdentifier&gt; instead of a bare FrameIdentifier. When
nullptr is passed (indicating focus has moved to chrome), it broadcasts
std::nullopt to other WebContent processes, allowing them to properly
clear their focused frame state via setFocusedFrame(nullptr).

Test: FocusWebView.CrossOriginIframeRelinquishToChromeTests

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::focusedFrameChanged):
(WebKit::WebPageProxy::broadcastFocusedFrameToOtherProcesses):
(WebKit::WebPageProxy::focusRemoteFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::frameWasFocusedInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
(TestWebKitAPI::CrossOriginIframeRelinquishToChromeTests::runTest)
Parameterized test that verifies document.activeElement is correctly
cleared when focus moves from a cross-origin iframe to an AppKit text
field (emulating some user agent chrome), with and without site isolation
enabled.

Canonical link: <a href="https://commits.webkit.org/301071@main">https://commits.webkit.org/301071@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe79ff27bce2aa39783179aa881278b8f6ffc3fc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124863 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44534 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35270 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131709 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76766 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/780b294b-46e5-4d31-a590-a3fe963ff12f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126740 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45236 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53103 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95020 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63045 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f90a8c7c-1646-43b5-a606-d1cc6c07d415) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127817 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36093 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111670 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75577 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1bc4c3f5-13fb-4500-82df-4961f75730b7) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35026 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29827 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75185 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105853 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30055 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134377 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51705 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39510 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103500 "11 flakes 54 failures") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52122 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107885 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103262 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48622 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26909 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48724 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19567 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51581 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57389 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50970 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54326 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52664 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
```diff
@@ -196,6 +213,120 @@ HTTPServer server({
+    Util::run(&takeFocusCalled);
+    checkFocusState({ "mainFrameBody", false }, { "iframeBody", false });
+    EXPECT_TRUE(textField.get().didBecomeFirstResponder);
+    EXPECT_FALSE(textField.get().didSeeKeyDownEvent);
```
Nit:
```suggestion
    EXPECT_TRUE([textField didBecomeFirstResponder]);
    EXPECT_FALSE([textField didSeeKeyDownEvent]);
```

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/fe79ff27bce2aa39783179aa881278b8f6ffc3fc

aprotyas committed:
[Site Isolation] Main frame's Document.activeElement incorrectly reports iframe after focus is relinquished from a cross-origin iframe to the user agent chrome
https://bugs.webkit.org/show_bug.cgi?id=300154
rdar://161863002

Reviewed by Aditya Keerthi.

When focus is relinquished from a cross-origin iframe to the user agent
chrome (e.g., an AppKit text field), document.activeElement in the main
frame incorrectly still references the iframe element. This occurs
because broadcastFocusedFrameToOtherProcesses() only supported
broadcasting a valid FrameIdentifier, not nullptr when focus moves to
chrome.

The fix changes broadcastFocusedFrameToOtherProcesses() to accept
std::optional<FrameIdentifier> instead of a bare FrameIdentifier. When
nullptr is passed (indicating focus has moved to chrome), it broadcasts
std::nullopt to other WebContent processes, allowing them to properly
clear their focused frame state via setFocusedFrame(nullptr).

Test: FocusWebView.CrossOriginIframeRelinquishToChromeTests

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::focusedFrameChanged):
(WebKit::WebPageProxy::broadcastFocusedFrameToOtherProcesses):
(WebKit::WebPageProxy::focusRemoteFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::frameWasFocusedInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
(TestWebKitAPI::CrossOriginIframeRelinquishToChromeTests::runTest)
Parameterized test that verifies document.activeElement is correctly
cleared when focus moves from a cross-origin iframe to an AppKit text
field (emulating some user agent chrome), with and without site isolation
enabled.

Canonical link: https://commits.webkit.org/301071@main
